GraalVM-Truffle.md

```diff
@@ -1,3 +1,116 @@
+- It provides a runtime environment in which Java bytecode can be executed.
+
+### JVM Bytecode
+
+- JVM bytecode is the intermediate representation of a Java program that is executed by the JVM.
+- It is generated by the Java compiler from the source code.
+
+### Ahead-of-Time (AOT) Compilation
+
+- Ahead-of-Time (AOT) compilation is a process where the source code is compiled into machine code before execution.
+- The resulting machine code can be executed directly by the operating system without the need for a JVM [4].
+
+### Just-In-Time Compiler (JIT)
+
+- A Just-In-Time (JIT) compiler is a component of the JVM.
+- Compiles sections of Java bytecode at interpreter runtime [4,7,8]
+
+## Truffle Language Implementation
+
+The Truffle Framework allows developers to create interpreters for guest languages that run on GraalVM. You can compile the Guest Language Applications with the Guest Language Implementation to JVM bytecode [2].
+
+The FMC diagram visualizes the Truffle language implementation processing.
+
+![Language Implementation Processing](img/graal-fmc-Truffle.drawio.png)
+
+### Truffle Framework
+
+- Truffle is a framework, which provides APIs to create language interpreters and compilers.
+- The resulting applications run on the GraalVM [1,2,5].
+
+### Guest Language Implementation
+
+- The Guest Language Implementation is the code that defines the semantics and behavior of a specific programming language.
+- It is built using the Truffle APIs.
+
+### Guest Language Interpreter
+
+- The Guest Language Interpreter is the runtime component that executes the guest language code.
+- It is compiled with a Java compiler from the Guest Language Implementation and the Truffle Framework.
+- Given a Guest Language Application, the Guest Language Interpreter executes the code and provides the necessary runtime support for the language features.
+  
+### Guest Language Application
+
+- The Guest Language Application is the code written in the guest language that is executed by the Guest Language Interpreter.
+- It can be any program or script that uses the features of the guest language.
+
+### HotSpotVM and GraalVM
+
+- The HotSpotVM is the Java Virtual Machine that executes the Guest Language Application.
+- It uses the GraalVM as a JIT compiler to optimize the execution of the guest language code.
+- The GraalVM applies advanced and speculative optimizations to improve performance [1,5].
+
+## GraalVM Compilation Process
+
+Given a JVM bytecode, you can use GraalVM to run and optimize it. The FMC diagram visualizes the JVM mode of GraalVM.
+
+![GraalVM Architecture](img/graal-fmc-Graal.drawio.png)
+
+### HotSpotVM
+
+- The HotSpotVM works as Java Virtual Machine (JVM) and executes Java applications using tiered compilation.
+- Initially, the interpreter runs the bytecode. As the code becomes "hot," it is compiled first with the client compiler (C1) and later with the server compiler (C2) [7].
```

Comment:
I would rather say that the GraalVM is a JVM that is based on the HotSpotVM and switches out the Compilers used of it (C1/C2) with the Graal Compiler that supports tiered compilation (as an analogy to C1/C2) and applis aggressive and speculative optimizations. Also GraalVM != Graal Compiler, no?

Reply:
I completely agree, and I'll edit the GraalVM section. However, HotSpotVM, as it is, is also a JVM we could normally use, right? I tried to introduce it as a standalone component. Is it too complex?

> Also GraalVM != Graal Compiler, no?

Yes!

Reply:
>However, HotSpotVM, as it is, is also a JVM we could normally use, right?
Yes, its possible to execute all Truffle languages on a regular JVM such as the HotSpotVM. But I'm not sure how specify this. Maybe more like "regular Java applications" / "not Guest language application specific"? 🤔

Reply:
LGTM! 👍